src/metalookup/core/content.py

```diff
@@ -0,0 +1,174 @@
+import asyncio
+import json
+import logging
+import re
+from asyncio import Task
+from typing import Optional
+
+from aiohttp import ClientSession
+from bs4 import BeautifulSoup
+from pydantic import HttpUrl
+from tldextract.tldextract import TLDExtract
+
+from metalookup.app.splash_models import HAR, SplashResponse
+from metalookup.lib.constants import SCRIPT
+from metalookup.lib.settings import SPLASH_HEADERS, SPLASH_TIMEOUT, SPLASH_URL
+from metalookup.lib.tools import get_unique_list, runtime
+
+logger = logging.getLogger(__file__)
+
+
+class Content:
+    """
+    Enrich the URL of a website with it's html content, HAR, and other features.
+
+    Since the provided features potentially rely on each other, and the html body and HAR are acquired asynchronously,
+    all provided features are exposed via async functions. However, under the hood, the instance caches these features
+    such that only the first call is potentially expensive.
+    This will also allow to move some the computation of some expensive features into a dedicated process pool to
+    mitigate blocking the async runtime for to long.
+    """
+
+    tld_extractor: TLDExtract = TLDExtract(cache_dir=None)
+
+    def __init__(self, url: HttpUrl, html: Optional[str] = None, har: Optional[HAR] = None):
+        self.url = url
+        self._html: Optional[str] = html
+        self._har: Optional[HAR] = har
+        self._task: Optional[Task[SplashResponse]] = None
+        self._domain: Optional[str] = None
+        self._soup: Optional[BeautifulSoup] = None
+        self._headers: Optional[dict[str, str]] = None
+        self._raw_links: Optional[list[str]] = None
+
+    async def _fetch(self) -> tuple[str, HAR]:
+        # a pretty primitive implementation of request deduplication.
+        # If a request is issued, we check if there is already a request for
+        # given URL (self._task).
+        # - If so, we simply await the response of that request instead of
+        #   issuing a new one (that's why tasks are used of plain
+        #   awaitables - one can await a task multiple times).
+        # - If not, then a respective task is created.
+        assert self._html is None, "Should not fetch when html is already preset"
```

Comment:
Good point. That would be an `AssertionError` holding the `"Should not fetch when html is already preset"` message. This exception would bubble up from the extractor that causes it, through the `MetadataManager` (the `await gather(*tasks)` raises exceptions of tasks to where one awaits the task). From there the exception would bubble up into the apilayer/fasapi library where it would be turned into an internal server error response.

Reply:
So the whole exception goes up until the API because - fetch was called at the wrong place? That would be if fetch fails to set html or har, right? Basically, this is a doubling to the check you have before calling fetch, right? At least in the current implementation

Reply:
Yep, `_fetch` is private and should never be called from the wrong place :) Hence the assertions!

Reply:
then why not double_under __ it, so the name get's wrangled? That is safer than the assertions.